File: back/src/main/java/com/triptrace/domain/image/image/storage/ImageFileStorage.java
package com.triptrace.domain.image.image.storage;

import java.awt.image.BufferedImage;
import java.io.IOException;
import java.nio.file.Path;
import java.util.UUID;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import com.triptrace.domain.image.image.error.ImageErrorCode;
import com.triptrace.domain.image.image.processing.ExifOrientation;
import com.triptrace.domain.image.image.processing.ImageProcessor;
import com.triptrace.domain.image.image.processing.dto.SavedFileInfo;
import com.triptrace.domain.image.image.processing.dto.StoredFile;
import com.triptrace.domain.image.image.exception.ImageProcessException;

@Component
public class ImageFileStorage {
    @java.lang.SuppressWarnings("all")
    private static final org.slf4j.Logger log = org.slf4j.LoggerFactory.getLogger(ImageFileStorage.class);
    private final String uploadDir;
    private final String profileImagesPath;
    private final String servingImagesPath;
    private final String thumbnailImagesPath;
    private final String publicPrefix;
    private final int thumbnailWidth;
    private final int thumbnailHeight;
    private final String jpegExt;
    private final FileStorage fileStorage;
    private final ImageProcessor imageProcessor;

    public ImageFileStorage(ImageStorageProperties properties, FileStorage fileStorage) {
        this(properties, fileStorage, new ImageProcessor());
    }

    @Autowired
    public ImageFileStorage(ImageStorageProperties properties, FileStorage fileStorage, ImageProcessor imageProcessor) {
        this.uploadDir = properties.upload().path();
        this.servingImagesPath = properties.upload().servingPath();
        this.thumbnailImagesPath = properties.upload().thumbnailPath();
        this.thumbnailWidth = properties.thumbnail().width();
        this.thumbnailHeight = properties.thumbnail().height();
        this.jpegExt = properties.ext().jpg();
        this.profileImagesPath = properties.upload().profilePath();
        this.publicPrefix = properties.upload().publicPrefix();
        this.fileStorage = fileStorage;
        this.imageProcessor = imageProcessor;
    }

    public String saveProfileImage(byte[] image) throws ImageProcessException {
        BufferedImage bufferedImage = imageProcessor.read(image);
        StoredFile stored = saveImage(
            bufferedImage,
            resolveUploadPath(profileImagesPath),
            generateFileName(jpegExt),
            false
        );
        return profileImagesPath + "/" + stored.name();
    }

    private StoredFile saveImage(
        BufferedImage image,
        String directoryPath,
        String fileName,
        boolean isThumbnail)
        throws ImageProcessException {
        StoredFile storedFile = null;
        try {
            if (isThumbnail) {
                image = imageProcessor.resizeToFit(image, thumbnailWidth, thumbnailHeight);
            }
            byte[] imageBytes = imageProcessor.encodeJpeg(image, jpegExt);
            storedFile = fileStorage.save(imageBytes, directoryPath, fileName);
        } catch (IOException e) {
            log.warn(e.getMessage());
            throw new ImageProcessException(ImageErrorCode.SAVE_ERROR);
        }
        if (storedFile == null) {
            throw new ImageProcessException(ImageErrorCode.SAVE_ERROR);
        }
        return storedFile;
    }

    public SavedFileInfo saveImageWithThumbnail(
        byte[] image,
        ExifOrientation orientation)
        throws ImageProcessException {
        BufferedImage bufferedImage = imageProcessor.read(image);
        bufferedImage = imageProcessor.rotate(bufferedImage, orientation);
        StoredFile origin = saveImage(
            bufferedImage,
            resolveUploadPath(servingImagesPath),
            generateFileName(jpegExt),
            false
        );
        StoredFile thumbnail;
        try {
            thumbnail = saveImage(
                bufferedImage,
                resolveUploadPath(thumbnailImagesPath),
                generateFileName(jpegExt),
                true);
        } catch (ImageProcessException e) {
            deleteImage(servingImagesPath + "/" + origin.name());
            throw e;
        }
        return new SavedFileInfo(
            servingImagesPath + "/" + origin.name(),
            thumbnailImagesPath + "/" + thumbnail.name(),
            origin.size(),
            "image/" + jpegExt
        );
    }

    public boolean deleteImage(String imagePath) throws ImageProcessException {
        try {
            fileStorage.delete(resolveStoragePath(imagePath));
        } catch (IOException e) {
            log.warn(imagePath, e);
            throw new ImageProcessException(ImageErrorCode.DELETE_ERROR);
        }
        return true;
    }

    private String generateFileName(String fileExt) {
        return UUID.randomUUID() + "." + fileExt;
    }

    private String resolveUploadPath(String path) {
        return Path.of(uploadDir, path.replaceFirst("^/", "")).toString();
    }

    private String resolveStoragePath(String imagePath) {
        return resolveUploadPath(imagePath);
    }

    public void cleanUp(SavedFileInfo savedFileInfo) {
        String originFile = savedFileInfo.servingUrl();
        String thumbnailFile = savedFileInfo.thumbnailUrl();
        try {
            deleteImage(originFile);
        } catch (ImageProcessException e) {
            log.warn(e.getMessage());
            throw new ImageProcessException(ImageErrorCode.REWARD_TRANSACTION_ERROR);
        }
        try {
            deleteImage(thumbnailFile);
        } catch (ImageProcessException e) {
            log.warn(e.getMessage());
            throw new ImageProcessException(ImageErrorCode.REWARD_TRANSACTION_ERROR);
        }
    }
}

File: back/src/main/java/com/triptrace/domain/image/image/storage/ImageFileStorage.kt
package com.triptrace.domain.image.image.storage

import com.triptrace.domain.image.image.error.ImageErrorCode
import com.triptrace.domain.image.image.exception.ImageProcessException
import com.triptrace.domain.image.image.processing.ExifOrientation
import com.triptrace.domain.image.image.processing.ImageProcessor
import com.triptrace.domain.image.image.processing.dto.SavedFileInfo
import com.triptrace.domain.image.image.processing.dto.StoredFile
import com.triptrace.global.app.Domain
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.awt.image.BufferedImage
import java.io.IOException
import java.nio.file.Path
import java.util.UUID

@Component
class ImageFileStorage @Autowired constructor(
    properties: ImageStorageProperties,
    private val fileStorage: FileStorage,
    private val imageProcessor: ImageProcessor,
) {
    private val uploadDir = properties.upload().path()
    private val profileImagesPath = properties.upload().profilePath()
    private val servingImagesPath = properties.upload().servingPath()
    private val thumbnailImagesPath = properties.upload().thumbnailPath()
    private val thumbnailWidth = properties.thumbnail().width()
    private val thumbnailHeight = properties.thumbnail().height()
    private val jpegExt = properties.ext().jpg()

    constructor(properties: ImageStorageProperties, fileStorage: FileStorage) : this(
        properties,
        fileStorage,
        ImageProcessor(),
    )

    fun saveProfileImage(image: ByteArray?): String {
        val stored = saveImage(
            imageProcessor.read(image),
            resolveUploadPath(profileImagesPath),
            generateFileName(jpegExt),
            false,
        )
        return "$profileImagesPath/${stored.name()}"
    }

    fun saveImageWithThumbnail(image: ByteArray?, orientation: ExifOrientation?): SavedFileInfo {
        if (image == null) {
            throw ImageProcessException(ImageErrorCode.READ_ERROR)
        }
        val bufferedImage = imageProcessor.rotate(imageProcessor.read(image), orientation)
        val origin = saveImage(
            bufferedImage,
            resolveUploadPath(servingImagesPath),
            generateFileName(jpegExt),
            false,
        )
        val thumbnail = try {
            saveImage(
                bufferedImage,
                resolveUploadPath(thumbnailImagesPath),
                generateFileName(jpegExt),
                true,
            )
        } catch (exception: ImageProcessException) {
            deleteImage("$servingImagesPath/${origin.name()}")
            throw exception
        }
        return SavedFileInfo(
            "$servingImagesPath/${origin.name()}",
            "$thumbnailImagesPath/${thumbnail.name()}",
            origin.size(),
            "image/$jpegExt",
        )
    }

    fun deleteImage(imagePath: String?): Boolean {
        try {
            fileStorage.delete(resolveStoragePath(imagePath))
        } catch (exception: IOException) {
            log.warn(
                "[{}] image file storage image path: {} reason: {}",
                Domain.IMAGE.name,
                imagePath,
                exception.message,
            )
            throw ImageProcessException(ImageErrorCode.DELETE_ERROR)
        }
        return true
    }

    fun cleanUp(savedFileInfo: SavedFileInfo) {
        try {
            deleteImage(savedFileInfo.servingUrl())
        } catch (exception: ImageProcessException) {
            log.warn("[{}] image file storage fallback reason: {}", Domain.IMAGE.name, exception.message)
            throw ImageProcessException(ImageErrorCode.REWARD_TRANSACTION_ERROR)
        }

        try {
            deleteImage(savedFileInfo.thumbnailUrl())
        } catch (exception: ImageProcessException) {
            log.warn("[{}] image file storage fallback reason: {}", Domain.IMAGE.name, exception.message)
            throw ImageProcessException(ImageErrorCode.REWARD_TRANSACTION_ERROR)
        }
    }

    private fun saveImage(
        image: BufferedImage,
        directoryPath: String,
        fileName: String,
        isThumbnail: Boolean,
    ): StoredFile {
        try {
            val imageToSave = if (isThumbnail) {
                imageProcessor.resizeToFit(image, thumbnailWidth, thumbnailHeight)
            } else {
                image
            }
            return fileStorage.save(imageProcessor.encodeJpeg(imageToSave, jpegExt), directoryPath, fileName)
                ?: throw ImageProcessException(ImageErrorCode.SAVE_ERROR)
        } catch (exception: IOException) {
            log.warn("[{}] image processor fallback reason: {}", Domain.IMAGE.name, exception.message)
            throw ImageProcessException(ImageErrorCode.SAVE_ERROR)
        }
    }

    private fun generateFileName(fileExt: String) = "${UUID.randomUUID()}.$fileExt"

    private fun resolveUploadPath(path: String): String =
        Path.of(uploadDir, path.replaceFirst("^/", "")).toString()

    private fun resolveStoragePath(imagePath: String?) = resolveUploadPath(imagePath!!)

    private companion object {
        private val log = LoggerFactory.getLogger(ImageFileStorage::class.java)
    }
}
